velox/functions/sparksql/String.h

```diff
@@ -1181,4 +1181,56 @@ struct FindInSetFunction {
   }
 };
 
+template <typename T>
+struct SoundexFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  static constexpr bool is_default_ascii_behavior = true;
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  void call(out_type<Varchar>& result, const arg_type<Varchar>& input) {
+    size_t inputSize = input.size();
+    if (inputSize == 0) {
+      result.setEmpty();
+      return;
+    }
+    if (!std::isalpha(input.data()[0])) {
+      // First character must be a letter, otherwise input is returned.
+      result.setNoCopy(input);
+      return;
+    }
+    result.resize(4);
+    result.data()[0] = std::toupper(input.data()[0]);
+    int32_t sxi = 1;
+    int32_t idx = result.data()[0] - 'A';
+    char lastCode = kUSEnglishMapping[idx];
+    for (auto i = 1; i < inputSize; ++i) {
+      if (!std::isalpha(input.data()[i])) {
+        lastCode = '0';
+        continue;
+      }
+      idx = std::toupper(input.data()[i]) - 'A';
+      char code = kUSEnglishMapping[idx];
+      if (code != '7') {
+        if (code != '0' && code != lastCode) {
+          result.data()[sxi++] = code;
+          if (sxi > 3) {
+            break;
+          }
+        }
+        lastCode = code;
+      }
+    }
+    for (; sxi < 4; sxi++) {
+      result.data()[sxi] = '0';
+    }
+  }
+
+ private:
+  /// Soundex mapping table.
+  /// For details, please see https://en.wikipedia.org/wiki/Soundex.
```

Comment:
Perhaps, move this comment to the start of 'call'. The reader of this code needs to read this article before being able to understand the code.